gamification-evm-services/src/main/java/io/meeds/evm/gamification/service/EvmContractTransferService.java

```diff
@@ -43,93 +49,115 @@ public class EvmContractTransferService {
   @Autowired
   private RuleService           ruleService;
 
-  public void listenEvmContractTransfer(RuleDTO rule) {
+  @Autowired
+  private WalletAccountService  walletAccountService;
+
+  @Autowired
+  private SettingService        settingService;
+
+  public void scanForContractTransactions(RuleDTO rule) {
     String trigger = rule.getEvent().getTrigger();
     String blockchainNetwork = rule.getEvent().getProperties().get(Utils.BLOCKCHAIN_NETWORK);
     String contractAddress = rule.getEvent().getProperties().get(Utils.CONTRACT_ADDRESS).toLowerCase();
     Long networkId = Long.parseLong(rule.getEvent().getProperties().get(Utils.NETWORK_ID));
-    Long lastIdProcced = getLastIdProcced(rule, contractAddress, networkId);
-    List<EvmTransaction> transactions = evmTransactionService.getTransactionsByContractAddressAndNetworkIdFromId(contractAddress,
-                                                                                                                 networkId,
-                                                                                                                 lastIdProcced);
-    if (CollectionUtils.isNotEmpty(transactions)) {
-      transactions.forEach(transaction -> {
-        try {
-          handleEvmTrigger(rule, transaction, trigger, contractAddress, networkId, blockchainNetwork);
-        } catch (Exception e) {
-          LOG.warn("Error broadcasting EVM event for transaction {} and trigger {}",
-                   transaction.getTransactionHash(),
-                   trigger,
-                   e);
+    List<Wallet> wallets = getAllDistinctWalletsFromTransactions(contractAddress, rule.getCreatedDate());
```

Comment:
It seems that you need only Wallet Addresses not the whole `Wallet` object, WDYT ?

```suggestion
    List<String> addresses = getAllDistinctWalletsFromTransactions(contractAddress, rule.getCreatedDate());
```